pkg/ipmanager/ipam/controller.go

```diff
@@ -0,0 +1,236 @@
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			service, err := i.kubeClientSet.CoreV1().Services(ns).Get(name, metav1.GetOptions{})
+
+			if err != nil {
+				i.serviceWorkqueue.Forget(obj)
+				return fmt.Errorf("error retrieving submariner-ipam-controller service object %s: %v", name, err)
+			}
+			annotations := service.GetAnnotations()
+			if annotations == nil {
+				annotations = map[string]string{}
+			}
+			if annotations[submarinerIpamGlobalIp] == "" {
+				annotations[submarinerIpamGlobalIp] = "169.254.1.1"
+				service.SetAnnotations(annotations)
+				klog.V(4).Infof("Updating service %s", service.Name)
+				_, updateErr := i.kubeClientSet.CoreV1().Services(ns).Update(service)
+				return updateErr
+			}
+			return nil
+		})
+		if retryErr != nil {
+			i.serviceWorkqueue.Forget(obj)
+			return retryErr
+		}
+
+		i.serviceWorkqueue.Forget(obj)
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (i *Controller) processNextPod() bool {
+	obj, shutdown := i.podWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer i.podWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			i.podWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+		if i.excludeNamespaces[ns] {
+			i.podWorkqueue.Forget(obj)
+			return nil
+		}
+
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of Pod before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			pod, err := i.kubeClientSet.CoreV1().Pods(ns).Get(name, metav1.GetOptions{})
+
+			if err != nil {
+				i.podWorkqueue.Forget(obj)
+				return fmt.Errorf("error retrieving submariner-ipam-controller pod object %s: %v", name, err)
+			}
+			annotations := pod.GetAnnotations()
+			if annotations == nil {
+				annotations = map[string]string{}
+			}
+			if annotations[submarinerIpamGlobalIp] == "" {
+				annotations[submarinerIpamGlobalIp] = "169.254.1.1"
+				pod.SetAnnotations(annotations)
+				klog.V(4).Infof("Updating pod %s", pod.Name)
+				_, updateErr := i.kubeClientSet.CoreV1().Pods(ns).Update(pod)
+				return updateErr
+			}
+			return nil
+		})
+		if retryErr != nil {
+			i.podWorkqueue.Forget(obj)
+			return retryErr
+		}
+
+		i.podWorkqueue.Forget(obj)
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (i *Controller) enqueueService(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	klog.V(6).Infof("Enqueueing service for ipam controller %v", obj)
+	i.serviceWorkqueue.AddRateLimited(key)
+}
+
+func (i *Controller) enqueuePod(obj interface{}) {
```

Comment:
This method (including enqueuService) gets invoked for every 60 seconds. I've seen this as a pattern in the current code. I have a feeling that this is not scalable. Imagine we have 10,000 PODs in a cluster. We get an event for every POD and that gets repeated for every 60 secs. @tpantelis @mpeterson your thoughts on this?

Reply:
I agree - we shouldn't need such a frequent resync period. If we handle retries then I think we can set the resync period to 0.

Reply:
What would resync 0 mean? What will happen if it failed and we return error when processing the request? Will it requeue?

Reply:
On error when processing a resource from the queue, it's up to us to re-queue it if we deem it a transient error such that a retry might succeed.  I believe a 0 resync period disables unsolicited periodic resyncing but it will still resync on reconnection to the backend.  A periodic resync is basically a fallback to re-process resources that might have previously failed. It will send an update for every resource even if nothing changed which, as @sridhargaddam points out, may be problematic when there's lot of resources, especially with a frequent period like 60 sec. That frequency is probably OK for the subm components that process Endpoint and Cluster resources as there's not a lot of those (essentially one per cluster). In this case, maybe a less frequent resync period of say every hour would be reasonable - also maybe we could shortcut the processing on update by checking if the pod/service already has the annotation before queueing it.